mm2src/coins/lightning/ln_platform.rs

```diff
@@ -0,0 +1,583 @@
+        registered_outputs.push(output);
+    }
+
+    async fn get_tx_if_onchain(&self, txid: Txid) -> Result<Option<Transaction>, GetTxError> {
+        match self
+            .coin
+            .as_ref()
+            .rpc_client
+            .get_transaction_bytes(&H256Json::from(txid.as_hash().into_inner()).reversed())
+            .compat()
+            .await
+            .map_err(|e| e.into_inner())
+        {
+            Ok(bytes) => Ok(Some(deserialize(&bytes.into_vec())?)),
+            Err(err) => {
+                if let UtxoRpcError::ResponseParseError(ref json_err) = err {
+                    if let JsonRpcErrorType::Response(_, json) = &json_err.error {
+                        if let Some(message) = json["message"].as_str() {
+                            if message.contains("'code': -5") {
+                                return Ok(None);
+                            }
+                        }
+                    }
+                }
+                Err(err.into())
+            },
+        }
+    }
+
+    async fn process_tx_for_unconfirmation<T>(&self, txid: Txid, monitor: Arc<T>)
+    where
+        T: Confirm,
+    {
+        match self.get_tx_if_onchain(txid).await {
+            Ok(Some(_)) => {},
+            Ok(None) => {
+                info!(
+                    "Transaction {} is not found on chain. The transaction will be re-broadcasted.",
+                    txid,
+                );
+                monitor.transaction_unconfirmed(&txid);
+            },
+            Err(e) => error!(
+                "Error while trying to check if the transaction {} is discarded or not :{:?}",
+                txid, e
+            ),
+        }
+    }
+
+    pub async fn process_txs_unconfirmations(
+        &self,
+        chain_monitor: Arc<ChainMonitor>,
+        channel_manager: Arc<ChannelManager>,
+    ) {
+        // Retrieve channel manager transaction IDs to check the chain for un-confirmations
+        let channel_manager_relevant_txids = channel_manager.get_relevant_txids();
+        for txid in channel_manager_relevant_txids {
+            self.process_tx_for_unconfirmation(txid, channel_manager.clone()).await;
+        }
+
+        // Retrieve chain monitor transaction IDs to check the chain for un-confirmations
+        let chain_monitor_relevant_txids = chain_monitor.get_relevant_txids();
+        for txid in chain_monitor_relevant_txids {
+            self.process_tx_for_unconfirmation(txid, chain_monitor.clone()).await;
+        }
+    }
+
+    async fn get_confirmed_registered_txs(&self, client: &ElectrumClient) -> Vec<ConfirmedTransactionInfo> {
+        let registered_txs = self.registered_txs.lock().clone();
+        let mut confirmed_registered_txs = Vec::new();
+        for (txid, scripts) in registered_txs {
+            if let Some(transaction) =
+                ok_or_continue_after_sleep!(self.get_tx_if_onchain(txid).await, TRY_LOOP_INTERVAL)
```

Comment:
My thought process was that if there was an error it would mostly be because there is a problem connecting to electrums (internet connection down, electrums down, etc..), and waiting a bit might be better than continuing before reconnecting. 

If we didn't wait, this might lead to passing through all the registered_txs transactions in this loop without confirming any and waiting for the next block to confirm it, I actually encountered this case while testing when my internet went down for a few minutes. 

I realize now that there should be a big change in the confirmation process of registered_txs and the watching for spending of registered_outputs and that I should revise this whole process. I have a few ideas to solve this problem but I think this should be implemented in another PR.

One of these ideas is once a transaction or an output is registered through the register functions https://github.com/KomodoPlatform/atomicDEX-API/blob/9b3339a7e6e2e49ae98345de6162b4d6c31cbfed/mm2src/coins/lightning/ln_platform.rs#L527 https://github.com/KomodoPlatform/atomicDEX-API/blob/9b3339a7e6e2e49ae98345de6162b4d6c31cbfed/mm2src/coins/lightning/ln_platform.rs#L530 I should just spawn a separate thread for this transaction/output while retrying each step that involves calls to electrums until confirmation. One problem with that approach is this error for instance https://github.com/KomodoPlatform/atomicDEX-API/pull/1240#discussion_r852166780 where if a call fails for a reason unrelated to electrum connections, knowing that a certain output is spent might not resolve due to getting stuck in retrying such calls forever. This watched output might be related to an old commitment transaction broadcasted by a channel's malicious counterparty and we will not be able to broadcast a transaction to punish this counterparty for broadcasting an old commitment transaction without detecting the spending of the channel opening transaction by this old commitment transaction, what makes this worse is that there is a window for broadcasting this punishing transaction and if passed some of our funds will be stolen.

These problems can all be solved after watchtowers are implemented since watchtowers should be running a full node and if a call to electrums fails for whatever reason we can delegate the watching for a certain transaction or a spending of an output to a number of watchtowers while providing them with the transactions that should be broadcasted in such cases.

I think the left unresolved comments are related to the above problem and should be resolved in a new PR since this code might be changed. The `ok_or_retry_after_sleep` was used for instance in here https://github.com/KomodoPlatform/atomicDEX-API/pull/1240#discussion_r852166780 to avoid the above situation.

Reply:
The idea with the spawned thread or just a future sounds good! This will allow us to avoid infinite loop and to be sure that we'll try to broadcast transactions until they're accepted.
Just only one thing - I think that one separated thread is enough for this

Reply:
> Just only one thing - I think that one separated thread is enough for this :D

Big lightning nodes can have a lot of open channels, the most connected nodes have over 2000 open channels with them, which in turn means that they can watch for more than 2000 outputs if they are spent or not with every new block. Of course, these nodes should always use a full node as their chain source and not electrums, but I think it's the right way to at least spawn a thread for multiple transactions at a time but not all so as not to take a long time to finish the requests when a new block connects if a user who uses electrum decided to open a lot of channels.